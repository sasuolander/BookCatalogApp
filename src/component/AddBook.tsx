import React from 'react'

export const AddBook =({})=>{
    return(<div><p>test</p><form></form></div>)
}