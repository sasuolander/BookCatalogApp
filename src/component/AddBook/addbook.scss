.add-book {
  display: flex;

  flex-direction: column; // vertical layout
    padding: 10px 0 0 0;
  @media screen and (min-width: 480px) {
    flex-direction: row; // horizontal layout
  }
}

.form-group > * {
  flex: 1; // grow to fill
}

.form-group .label {
  flex: none; // don't grow
  @media screen and (min-width: 480px) {
    order: -1; // make sure the label is first
  }
}

.form-container .label {
  margin-right: 1em;
  font-weight: bold;
  @media screen and (min-width: 480px) {
    width: 150px;
    text-align: right;
  }
}

.form-container .input {
  width: 50%;
}

.form-container form {
  max-width: 640px;
  margin: 1em auto;
  padding: 1em;
}

.add-book-btn{
   

}
.container-btn{
    text-align: left;
    padding: 1em 0 0 0em ;
    margin-left: 164px; padding-left: 0em;
}
