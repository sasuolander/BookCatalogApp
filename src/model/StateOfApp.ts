export interface BookModel {
  author: String;
  title: String;
  itemType: String;
  abstract: String;
  date?: Date;
  isbn?: Number;
  DOI?: String;
  pages?: Number;
}

export interface ListOfBookState extends StatusOfAction {
  books: BookModel[];
}

export interface StatusOfAction {
  hasErrored?: Boolean;
}
