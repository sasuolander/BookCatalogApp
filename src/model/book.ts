export interface BookState{
    author:String,
    title:String,
    itemType:String,
    abstract:String,
    date?:Date,
    isbn?:Number,
    DOI?:String,
    pages?:Number
    }