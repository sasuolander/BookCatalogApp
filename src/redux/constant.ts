//Add
export enum ADD_BOOK {
     SUCCESS = "add_book/success",
     FAIL = "add_book/fail",
     PENDING = "add_book/pending"
}


//Loading book to list
export enum LOAD_BOOK {
    SUCCESS = "load_book/success",
    FAIL = "load_book/fail",
    PENDING = "load_book/pending",
}


//Edit book
export enum EDIT_BOOK {
    SUCCESS = "edit_book/success",
    FAIL = "edit_book/fail",
    PENDING = "edit_book/pending",
    PARTIAL = "edit_book/partial",
}
