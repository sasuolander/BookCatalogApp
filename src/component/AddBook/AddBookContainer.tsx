import React from "react";
import { Field, reduxForm } from "redux-form";
import { connect } from "react-redux";
import { AddBook } from "./AddBook";

interface formProps {}

const ConnectedAddBook = reduxForm<formProps>({ form: "formdemo" })(AddBook);

const mapStateToProps = (state: any, ownProps: any) => {
  return {
    prop: state.prop
  };
};

export default connect<{}, {}>(mapStateToProps)(ConnectedAddBook);
