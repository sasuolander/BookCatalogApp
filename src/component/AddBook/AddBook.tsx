import React, { FunctionComponent } from "react";
import { Field, InjectedFormProps } from "redux-form";
import { BookModel } from "./../../model/StateOfApp";

export interface componentAddProps extends InjectedFormProps {
  books?: [];
  addBookFunction?: Function;
}

export const AddBook: FunctionComponent<componentAddProps> = ({
  handleSubmit,
  addBookFunction
}): JSX.Element => {
  const createBookObject = (data: any): BookModel => {
    const book: BookModel = {
      title: data.title,
      author: data.author,
      abstract: data.abstract,
      itemType: data.itemType
    };
    return book;
  };

  const addToBooks = (book: BookModel): void => {
    addBookFunction!(book);
  };

  const createBookAndAddToBooks = (data: any): void => {
    addToBooks(createBookObject(data));
  };

  return (
    <div className={"form-container"}>
      <p>test</p>
      <form>
        <div className={"add-book"}>
          <label className={"label title"}>Title</label>
          <Field
            className={"input type"}
            name={"title"}
            type={"text"}
            component="input"
          />
        </div>

        <div className={"add-book"}>
          <label className={"label author"}>Author</label>
          <Field
            className={"input type"}
            name={"author"}
            type={"text"}
            component="input"
          />
        </div>

        <div className={"add-book"}>
          <label className={"label abstract"}>Abstract</label>
          <Field
            className={"input type"}
            name={"abstract"}
            type={"text"}
            component="input"
          />
        </div>

        <div className={"add-book"}>
          <label className={"label type"}>Type</label>
          <Field
            className={"input type"}
            name={"itemType"}
            type={"text"}
            component="input"
          />
        </div>
        <div className={"container-btn"}>
          <button
            type={"submit"}
            className={"add-book-btn"}
            onClick={handleSubmit(data => {
              createBookAndAddToBooks(data);
            })}
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  );
};
