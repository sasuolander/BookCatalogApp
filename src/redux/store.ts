import { applyMiddleware, createStore, compose } from "redux";
import thunk from "redux-thunk";
import rootReducer from "./reducer/mainReducer";

const middleware = [thunk];

const store = createStore(
  rootReducer,
  compose(
    applyMiddleware(...middleware),
    ((window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ &&
      (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__()) ||
      compose
  )
);

export default store;
