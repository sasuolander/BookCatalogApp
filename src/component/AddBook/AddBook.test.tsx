import React from "react";
import { AddBook, componentAddProps } from "./AddBook";
import { shallow } from "enzyme";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { combineReducers } from "redux";
import { reducer as formReducer } from "redux-form";
import { BookModel } from "./../../model/StateOfApp";

const rootReducer = combineReducers({
  form: formReducer
});

describe("Add book", () => {
  let wrap: any;
  const middlewares = [thunk];
  const mockStore = configureStore(middlewares);
  let store, formComponentWrapper;

  // @ts-ignore
  const defaultProps: componentAddProps = {
    handleSubmit() {
      return;
    },
    books: [],
    addBookFunction() {
      return;
    }
  };

  beforeEach(() => {
    const initialState = {};
    store = mockStore(initialState);

    wrap = shallow(<AddBook {...defaultProps} />);
  });
  it("render empty component Add book", () => {
    expect(wrap).toMatchSnapshot();
  });

  it("find a form element", () => {
    wrap.find("form");
  });

  it("find all non-optional fields in the form", () => {
    let fields: string[] = [];

    const book: BookModel = {
      author: "tim",
      title: "",
      abstract: "",
      itemType: ""
    };
    const list = Object.keys(book);

    wrap
      .find("form")
      .find(".add-book")
      .children(".input")
      .map((node: any) => {
        fields.push(node.props().name);
      });
    expect(fields).toEqual(expect.arrayContaining(list));
  });

  it("find all non-optionan labels in the form", () => {
    let labels: string[] = [];
    const expectedLabels: string[] = ["Title", "Author", "Abstract", "Type"];

    wrap
      .find("form")
      .find(".add-book")
      .children(".label")
      .map((node: any) => {
        labels.push(node.props().children);
      });

    expect(labels).toEqual(expect.arrayContaining(expectedLabels));
  });
});
