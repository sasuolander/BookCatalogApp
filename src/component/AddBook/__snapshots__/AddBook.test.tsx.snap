// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add book render empty component Add book 1`] = `
<div
  className="form-container"
>
  <p>
    test
  </p>
  <form>
    <div
      className="add-book"
    >
      <label
        className="label title"
      >
        Title
      </label>
      <Field
        className="input type"
        component="input"
        name="title"
        type="text"
      />
    </div>
    <div
      className="add-book"
    >
      <label
        className="label author"
      >
        Author
      </label>
      <Field
        className="input type"
        component="input"
        name="author"
        type="text"
      />
    </div>
    <div
      className="add-book"
    >
      <label
        className="label abstract"
      >
        Abstract
      </label>
      <Field
        className="input type"
        component="input"
        name="abstract"
        type="text"
      />
    </div>
    <div
      className="add-book"
    >
      <label
        className="label type"
      >
        Type
      </label>
      <Field
        className="input type"
        component="input"
        name="itemType"
        type="text"
      />
    </div>
    <div
      className="container-btn"
    >
      <button
        className="add-book-btn"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
